### PR TITLE
test(ws3): decouple sampling from logprob scoring

### LINKED ISSUE #135: [WS3] Sampling vs logprob decoupling
Part of #83 · Deferred to next phase — not in this month's sprint. Closely related to the cross-benchmark tool (#106 / WS3-5).

Rollout samples (temperature / top-p) while training computes teacher-forcing logprob. To compare apples to apples, the alignment target must be "logprob given the same token sequence" — sampling and logprob computation must be explicitly separated so the cross-engine comparison measures the same quantity, not the effect of sampling randomness.

Why this matters: if sampling is entangled with the logprob path, the cross-benchmark will show "drift" that is actually sampling variance, not operator/engine mismatch — a false positive that wastes debugging time.

Planned PRs:
- [ ] Define the comparison contract: both engines score the SAME fixed token sequence (teacher-forcing), independent of how the sequence was sampled.
- [ ] Decouple the rollout sampling step from the logprob-scoring step so the scored logprob is reproducible given a fixed sequence.
- [ ] Test: same token sequence → same logprob regardless of sampling params (temperature / top-p); confirm sampling variance is excluded from the drift measurement.
- [ ] Wire this contract into the cross-benchmark tool (#106 / WS3-5) so it always compares teacher-forcing logprobs.

### PR DESCRIPTION
**Pull request by haoruilee**
## Summary

- Add `teacher_forced_logprobs_reference` as the explicit reference path for scoring fixed token sequences.
- Mark rollout-derived training batches with `logprob_scoring_mode=teacher_forcing` so sampling params stay scoped to generation.
- Update the DeepSpeed worker and single-GPU GRPO example to use the teacher-forced scoring helper.
- Add tests proving the same sampled tokens score identically under different temperature/top-p/top-k settings, while changed token sequences change the score.
- Document the teacher-forced logprob testing contract.

Closes #135

## Validation

- `python3 -m pytest tests/test_reference_ops.py tests/test_training_contract.py -v`
- `python3 -m pytest tests/test_reference_ops.py tests/test_training_contract.py tests/test_deepspeed_training_worker.py tests/test_grpo_single_gpu_example.py -q`
- `/tmp/rl-kernel-test-venv/bin/python -m pytest tests/ rl_engine/tests/ -q`
- `uvx ruff check rl_engine/testing/reference_ops.py rl_engine/testing/__init__.py rl_engine/executors/training_contract.py rl_engine/executors/deepspeed_trainer.py examples/grpo_single_gpu.py tests/test_reference_ops.py tests/test_training_contract.py`
- `uvx mypy --ignore-missing-imports rl_engine/`
- `uvx pre-commit run --files rl_engine/testing/reference_ops.py rl_engine/testing/__init__.py rl_engine/executors/training_contract.py rl_engine/executors/deepspeed_trainer.py examples/grpo_single_gpu.py tests/test_reference_ops.py tests/test_training_contract.py docs/contributing/testing.md`
- `/tmp/rl-kernel-docs-venv/bin/mkdocs build --strict -f mkdocs.yaml`
- `python3 examples/grpo_single_gpu.py --device cpu --steps 2 --num-prompts 1 --samples-per-prompt 2 --prompt-len 2 --completion-len 3 --vocab-size 16 --hidden-dim 8`

## Summary by CodeRabbit

* **Documentation**
  * Updated testing guidance with instructions on verifying teacher-forced log-probability checks for consistency and benchmarking.

* **Tests**
  * Added comprehensive test coverage for training batch scoring and reference operations to validate consistency across different configurations.

* **Improvements**
  * Updated log-probability computation methodology across training components for consistency.

### CODE REVIEW

**Comment by Flink-ddd** [comment]
Hi @haoruilee, Thank you for your contribution, but It has already been assigned, and you can claim the non-WS issue like kernel or benchmark and observability etc.

File: rl_engine/executors/training_contract.py [hunk 1]
```diff
@@ -78,6 +78,7 @@ def _batch_from_rollout_or_synthetic(
         if token_groups:
             return self._batch_from_token_groups(token_groups, rollout), {
                 "training_data_source": "rollout_payload",
+                "logprob_scoring_mode": "teacher_forcing",
```
Comment (inaniloquentee) [inline rl_engine/executors/training_contract.py:81]: This should not be labeled teacher_forcing unless the rollout prompt context is preserved and used for scoring.

File: rl_engine/executors/training_contract.py
Comment (inaniloquentee) [inline rl_engine/executors/training_contract.py:126]: This only keeps completion tokens; prompt_token_ids from the rollout payload are dropped, so scoring loses the conditioning context.

Comment (inaniloquentee) [inline rl_engine/executors/training_contract.py:133]: Using zero prompt tokens makes different prompts indistinguishable when they share the same sampled completion.

File: rl_engine/executors/deepspeed_trainer.py [hunk 2]
```diff
@@ -106,7 +106,7 @@ def train(self, rollout: RolloutStageResult) -> TrainingStageResult:
         batch, payload_metrics = self._batch_from_rollout_or_synthetic(rollout)
 
         logits = _extract_logits(self.engine(batch.token_ids.long()))
```
Comment (inaniloquentee) [inline rl_engine/executors/deepspeed_trainer.py:108]: This forwards only completion token_ids, so current_logps are not computed as log p(completion | prompt).

File: tests/test_training_contract.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        ),
+    ]
+
+    assert [extract_rollout_token_groups(payload) for payload in payloads] == [
+        token_groups,
+        token_groups,
+    ]
+
+
+def test_same_sampled_tokens_score_the_same_under_different_sampling_params():
```
Comment (inaniloquentee) [inline tests/test_training_contract.py:59]: Please add a case with the same completion under different prompt_token_ids; those scores should differ under true teacher-forced scoring.

**Review by inaniloquentee — CHANGES_REQUESTED** [review]

**Comment by haoruilee** [comment]
Hi @a-kaa  @inaniloquentee ,

Thanks for the comments. As this issue has already been assigned, I will close this pr.
